### PR TITLE
Add Event Support to SignalFx Receiver

### PR DESCRIPTION
**Pull request by keitwb**
They are converted to logs in a special format.

### CODE REVIEW

File: receiver/signalfxreceiver/receiver.go [hunk 1]
```diff
@@ -90,27 +92,30 @@ type sfxReceiver struct {
...
+}
 
-	r := &sfxReceiver{
-		logger:       logger,
-		config:       &config,
-		nextConsumer: nextConsumer,
-	}
+func (r *sfxReceiver) RegisterMetricsConsumer(mc consumer.MetricsConsumer) {
+	r.Lock()
+	defer r.Unlock()
```
Comment (tigrannajaryan) [inline receiver/signalfxreceiver/receiver.go:109]: Is this function ever called concurrently? The lock doesn't hurt, but I wonder if this ever is needed just to be sure I understand everything correctly.

File: receiver/signalfxreceiver/receiver.go [hunk 2]
```diff
@@ -165,28 +175,21 @@ func (r *sfxReceiver) Shutdown(context.Context) error {
...
 
-func (r *sfxReceiver) handleReq(resp http.ResponseWriter, req *http.Request) {
-	transport := "http"
-	if r.config.TLSSetting != nil {
-		transport = "https"
-	}
-	ctx := obsreport.ReceiverContext(req.Context(), r.config.Name(), transport, r.config.Name())
-	ctx = obsreport.StartMetricsReceiveOp(ctx, r.config.Name(), transport)
-
+func (r *sfxReceiver) preRequest(ctx context.Context, resp http.ResponseWriter, req *http.Request) ([]byte, bool) {
```
Comment (tigrannajaryan) [inline receiver/signalfxreceiver/receiver.go:178]: Please add a comment to explain what this func does and what the return values are. It is not clear from the signature. Perhaps a better func name would help.

File: receiver/signalfxreceiver/receiver.go [hunk 3]
```diff
@@ -195,18 +198,44 @@ func (r *sfxReceiver) handleReq(resp http.ResponseWriter, req *http.Request) {
...
 
 	body, err := ioutil.ReadAll(bodyReader)
 	if err != nil {
 		r.failRequest(ctx, resp, http.StatusBadRequest, errReadBodyRespBody, err)
+		return nil, false
+	}
+	return body, true
+}
+
+func (r *sfxReceiver) postRequest(ctx context.Context, resp http.ResponseWriter, err error) {
```
Comment (tigrannajaryan) [inline receiver/signalfxreceiver/receiver.go:213]: Same here.

File: receiver/signalfxreceiver/signalfxv2_event_to_logdata.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/common/splunk"
+)
+
+// SignalFxV2ToMetricsData converts SignalFx proto data points to
+// consumerdata.MetricsData. Returning the converted data and the number of
+// dropped time series.
+func SignalFxV2EventsToLogRecords(
```
Comment (tigrannajaryan) [inline receiver/signalfxreceiver/signalfxv2_event_to_logdata.go:28]: Do we need this function to be exporter?

File: receiver/signalfxreceiver/signalfxv2_event_to_logdata.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+	logger *zap.Logger,
+	events []*sfxpb.Event,
+) pdata.LogSlice {
+	lrs := pdata.NewLogSlice()
+
```
Comment (tigrannajaryan) [inline receiver/signalfxreceiver/signalfxv2_event_to_logdata.go:33]: Can we call lrs.Resize() here to reduce the number of allocations? See recommendation here: https://github.com/open-telemetry/opentelemetry-collector/blob/639b9a80ed46ebe1ad244fcf064ed1d9e341056a/consumer/pdata/generated_log.go#L495

File: receiver/signalfxreceiver/signalfxv2_event_to_logdata.go [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+			attrs.InsertNull(splunk.SFxEventCategoryKey)
+		}
+
+		for _, dim := range event.Dimensions {
+			attrs.InsertString(dim.Key, dim.Value)
+		}
+
+		if len(event.Properties) > 0 {
+			propMap := pdata.NewAttributeMap()
+			for _, prop := range event.Properties {
```
Comment (tigrannajaryan) [inline receiver/signalfxreceiver/signalfxv2_event_to_logdata.go:62]: Similarly, call `propMap.InitEmptyWithCapacity` to preallocate the map capacity.

File: receiver/signalfxreceiver/signalfxv2_event_to_logdata.go [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+				// individually.
+				if prop.Value.StrValue != nil {
+					propMap.InsertString(prop.Key, prop.Value.GetStrValue())
+				} else if prop.Value.IntValue != nil {
+					propMap.InsertInt(prop.Key, prop.Value.GetIntValue())
+				} else if prop.Value.DoubleValue != nil {
+					propMap.InsertDouble(prop.Key, prop.Value.GetDoubleValue())
+				} else if prop.Value.BoolValue != nil {
+					propMap.InsertBool(prop.Key, prop.Value.GetBoolValue())
+				}
```
Comment (tigrannajaryan) [inline receiver/signalfxreceiver/signalfxv2_event_to_logdata.go:73]: Is there an `else` possible here?

File: receiver/signalfxreceiver/signalfxv2_event_to_logdata.go [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+		lr.InitEmpty()
+
+		// The EventType field is the most logical "name" of the event.
+		lr.SetName(event.EventType)
+
+		// SignalFx timestamps are in millis so convert to nanos by multiplying
+		// by 1 million.
+		lr.SetTimestamp(pdata.TimestampUnixNano(event.Timestamp * 1e6))
+
+		attrs := lr.Attributes()
```
Comment (tigrannajaryan) [inline receiver/signalfxreceiver/signalfxv2_event_to_logdata.go:45]: Similarly `attrs.InitEmptyWithCapacity` to preallocate the map capacity.

File: receiver/signalfxreceiver/signalfxv2_event_to_logdata.go [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+// dropped time series.
+func SignalFxV2EventsToLogRecords(
+	logger *zap.Logger,
+	events []*sfxpb.Event,
+) pdata.LogSlice {
+	lrs := pdata.NewLogSlice()
+
+	for _, event := range events {
+		lr := pdata.NewLogRecord()
+		lr.InitEmpty()
```
Comment (tigrannajaryan) [inline receiver/signalfxreceiver/signalfxv2_event_to_logdata.go:36]: If `Resize` is used this becomes `lr := lrs.At(i)` and `Append` at the end of the loop is no longer needed.

File: receiver/signalfxreceiver/receiver.go (hunk 1 above)
Comment (keitwb) [reply receiver/signalfxreceiver/receiver.go:109]: The lock also helps if the create*Receiver methods are called in separate goroutines, even if not concurrently, to establish memory coherency.  I think from looking at the service core code that they are not, but I would rather assume as little as possible in the receiver unless it is something documented.

That being said, I would also synchronize the use of the global map in the factory as well to be consistent.

File: receiver/signalfxreceiver/signalfxv2_event_to_logdata.go (hunk 7 above)
Comment (keitwb) [reply receiver/signalfxreceiver/signalfxv2_event_to_logdata.go:81]: If there isn't any value set in protobuf, then I added an else to make it insert a null attribute value with the key.

File: receiver/signalfxreceiver/signalfxv2_event_to_logdata.go (hunk 4 above)
Comment (keitwb) [reply receiver/signalfxreceiver/signalfxv2_event_to_logdata.go]: No.  It is now unexported.

File: receiver/signalfxreceiver/receiver.go (hunk 2 above)
Comment (keitwb) [reply receiver/signalfxreceiver/receiver.go]: Renamed to `readBody`.

File: receiver/signalfxreceiver/receiver.go (hunk 3 above)
Comment (keitwb) [reply receiver/signalfxreceiver/receiver.go]: Renamed to `writeResponse`.

**Review by tigrannajaryan — APPROVED** [review]

**Pull request merged**